vsphere/stackstate_checks/vsphere/vsphere.py

```diff
@@ -989,22 +989,24 @@ def _vsphere_vms(self, content, domain="Unspecified", regexes=None, include_only
             labels = []
             if not self._is_excluded(c, regexes, include_only_marked):
                 hostname = c.name
-
-                if isinstance(c, vim.VirtualMachine):
-                    topology_tags["topo_type"] = VSPHERE_COMPONENT_TYPE.VM
-                    topology_tags["name"] = c.name
-                    topology_tags["datastore"] = c.datastore[0]._moId
-                    topology_tags["layer"] = TOPOLOGY_LAYERS.VM
-                    topology_tags["domain"] = domain
-
-                    sts_identifiers, labels = self.extract_tags("VirtualMachine", c._moId)
-                    topology_tags["identifiers"] = sts_identifiers
-
-                    add_label_pair(labels, "name", topology_tags["name"])
-                    add_label_pair(labels, "guestId", c.config.guestId)
-                    add_label_pair(labels, "guestFullName", c.config.guestFullName)
-                    add_label_pair(labels, "numCPU", c.config.hardware.numCPU)
-                    add_label_pair(labels, "memoryMB", c.config.hardware.memoryMB)
+                try:
+                    if isinstance(c, vim.VirtualMachine):
+                        topology_tags["topo_type"] = VSPHERE_COMPONENT_TYPE.VM
+                        topology_tags["name"] = c.name
+                        topology_tags["layer"] = TOPOLOGY_LAYERS.VM
+                        topology_tags["domain"] = domain
+
+                        sts_identifiers, labels = self.extract_tags("VirtualMachine", c._moId)
+                        topology_tags["identifiers"] = sts_identifiers
+
+                        topology_tags["datastore"] = c.datastore[0]._moId
+                        add_label_pair(labels, "name", topology_tags["name"])
+                        add_label_pair(labels, "guestId", c.config.guestId)
+                        add_label_pair(labels, "guestFullName", c.config.guestFullName)
+                        add_label_pair(labels, "numCPU", c.config.hardware.numCPU)
+                        add_label_pair(labels, "memoryMB", c.config.hardware.memoryMB)
+                except Exception as e:
+                    self.log.exception("Exception occurred during collecting metadata for VMs: {}".format(str(e)))
                 topology_tags["labels"] = labels
                 obj_list.append(dict(mor_type="vm", hostname=hostname, topo_tags=topology_tags))
```

Comment:
If I read the code correctly, we are still sending a component even if an exception occurs. Depending on where the exception happened (either when we were still determining the required parameters (id, type, name, etc) or after that) we may or may not have a complete component definition.

I suggest you separate the retrieval of mandatory fields (and handle the exception by not sending a component) from optional fields (and handle the exception by sending the component and metadata we have collected).